core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java

```diff
@@ -474,6 +481,98 @@ public void testSerialization() throws Exception {
         assertThat(((CancelAllocationCommand) (sCommands.commands().get(4))).allowPrimary(), equalTo(true));
     }
 
+    private static void assertRerouteExplanation(RerouteExplanation actualExplanation,
+                                                 AllocationCommand expectedCommand,
+                                                 Decision.Type expectedDecision,
+                                                 List<String> expectedWarningSubstrings) {
+
+        assertThat(actualExplanation.command(), equalTo(expectedCommand));
+        assertThat(actualExplanation.decisions().type(), equalTo(expectedDecision));
+        assertThat(actualExplanation.warnings().size(), equalTo(expectedWarningSubstrings.size()));
+        for (int i = 0; i < expectedWarningSubstrings.size(); i++) {
+            String actualWarning = actualExplanation.warnings().get(i);
+            String expectedSubstring = expectedWarningSubstrings.get(i);
+            assertThat(actualWarning, containsString(expectedSubstring));
+        }
+    }
+
+    public void testRerouteExplanations() {
+        final String indexName = "fake_index";
+
+        AllocationService allocation = createAllocationService(Settings.builder()
+            .put(EnableAllocationDecider.CLUSTER_ROUTING_ALLOCATION_ENABLE_SETTING.getKey(), "none")
+            .put(EnableAllocationDecider.CLUSTER_ROUTING_REBALANCE_ENABLE_SETTING.getKey(), "none")
+            .build());
+
+        MetaData metaData = MetaData.builder()
+            .put(IndexMetaData.builder(indexName)
+                .settings(settings(Version.CURRENT))
+                .numberOfShards(1)
+                .numberOfReplicas(1)
+                .putInSyncAllocationIds(0, Collections.singleton("fake_insync_id"))) // todo test without
+            .build();
+
+        RoutingTable routingTable = RoutingTable.builder()
+            .addAsRecovery(metaData.index(indexName))
+            .build();
+
+        ClusterState clusterState = ClusterState.builder(ClusterName.DEFAULT)
+            .metaData(metaData)
+            .routingTable(routingTable)
+            .build();
+        clusterState = ClusterState.builder(clusterState)
+            .nodes(DiscoveryNodes.builder()
+                .add(newNode("data1"))
+                .add(newNode("data2"))
+                .add(newNode("data3")))
+            .build();
+        clusterState = allocation.reroute(clusterState, "reroute");
+
+        AllocationCommand command;
+        String warning;
+        if (randomBoolean()) {
+            command = new AllocateEmptyPrimaryAllocationCommand(indexName, 0, "data1", true);
+            warning = "Allocating an empty primary";
+        } else {
+            command = new AllocateStalePrimaryAllocationCommand(indexName, 0, "data1", true);
+            warning = "Allocating a stale primary";
```

Comment:
My intuition was that this wouldn't work, because it's allocating a stale primary before any replicas have been allocated. It does seem to though; in another version of this I put the RerouteExplanation assertions into the existing tests in this class and the allocation seemed to succeed. I'm not really sure what in the setup here is setting [the recovery type for that shard to `EXISTING_STORE`](https://github.com/elastic/elasticsearch/blob/0f7d52df689cd0c38b327f5b1050c46b25a68763/core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateStalePrimaryAllocationCommand.java#L120)